src/main/kotlin/com/openlattice/data/integration/destinations/PostgresDestination.kt

```diff
@@ -21,30 +21,280 @@
+import com.openlattice.postgres.PostgresArrays
+import com.zaxxer.hikari.HikariDataSource
+import org.apache.olingo.commons.api.edm.EdmPrimitiveTypeKind
+import org.slf4j.LoggerFactory
+import java.security.InvalidParameterException
+import java.sql.Connection
+import java.sql.PreparedStatement
 import java.util.*
 
 /**
  *
  * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
  */
-class PostgresDestination : IntegrationDestination {
+class PostgresDestination(
+        private val entitySets: Map<UUID, EntitySet>,
+        private val entityTypes: Map<UUID, EntityType>,
+        private val propertyTypes: Map<UUID, PropertyType>,
+        private val hds: HikariDataSource
+) : IntegrationDestination {
+    companion object {
+        private val logger = LoggerFactory.getLogger(PostgresDestination::class.java)
+        private val mapper = ObjectMappers.newJsonMapper()
+    }
+
     override fun integrateEntities(
             data: Set<Entity>, entityKeyIds: Map<EntityKey, UUID>, updateTypes: Map<UUID, UpdateType>
     ): Long {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+
+        return hds.connection.use { connection ->
+
+            data
+                    .groupBy({ it.entitySetId }, { normalize(entityKeyIds, it) })
+                    .map { (entitySetId, entities) ->
+                        val entitySet = entitySets.getValue(entitySetId)
+
+                        val partitions = entitySet.partitions.toList()
+                        val partitionsVersion = entitySet.partitionsVersion
+
+                        val baseVersion = System.currentTimeMillis()
+                        val tombstoneVersion = -baseVersion
+                        val writeVersion = baseVersion + 1
+                        val relevantPropertyTypes = entityTypes
+                                .getValue(entitySets.getValue(entitySetId).entityTypeId)
+                                .properties
+                                .associateWith(propertyTypes::getValue)
+                        val propertyTypeIdsArr = PostgresArrays.createUuidArray(connection, relevantPropertyTypes.keys)
+                        val writeVersionArray = PostgresArrays.createLongArray(connection, writeVersion)
+
+                        entities.groupBy { getPartition(it.first, partitions) }
+                                .map { (partition, entityPairs) ->
+                                    val entityMap = entityPairs.toMap()
+                                    val entityKeyIdsArr = PostgresArrays.createUuidArray(connection, entityMap.keys)
+                                    val partitionsArr = PostgresArrays.createIntArray(
+                                            connection, entityMap.keys.map { getPartition(it, partitions) }
+                                    )
+
+                                    if (updateTypes.getValue(entitySetId) == UpdateType.Replace) {
```

Comment:
do we also need to account for UpdateType.PartialReplace?

Reply:
Implemented.